### PR TITLE
Divide county pull list among the audit boards

### PR DESCRIPTION
**Pull request by gws**
**Ballot division:**

At a high level, the existing ballot sequence was left alone, with the audit board division overlaid on top of that sequence. The division only needs to be done once, resulting in an _index_ (the position that audit board is starting from in the full sorted ballot sequence), and a _count_ (the length of that audit board's segment, starting from its _index_) for each audit board.

The algorithm for dividing ballots works as follows:

- Divide the number of ballots to audit evenly across the audit boards, using integer division
- Take the remainder ballots and spread them across the audit boards, in audit board order
- Shuffle the result so the remainder ballots are not always biased toward the first audit boards across rounds.

The end result is a list that looks like `[8, 9, 8, 8]`, whose values are the number of ballots assigned to each board, in board index order.

This information can then be used to generate a list of maps like:

`[{"index": 0, "count": 8}, {"index": 8, "count": 9}, {"index": 17, "count": 8} ...]`

This data structure is saved on the Round, and then further used to determine which CVR is next for a given audit board. That part works by finding the first CVR without a corresponding ACVR (adhering to the sorted ballot order!) for each audit board, and we're off to the races.

In the end, rather than receiving a single CVR ID for the audit board to audit, the client receives a list of them, and depending on which audit board is signed in, that CVR will be chosen for audit.

**Sorting:**

Aside from the behavioral change, there was some decent amount of cleanup possible by making Comparable classes implement Comparable rather than containing their own inner comparator classes.

The sorting should also be slightly faster, since we don't check every attribute on every comparison unless it's needed for the comparison.

**References:**

- https://www.pivotaltracker.com/story/show/160019853

### CODE REVIEW

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRToAuditDownload.java [hunk 1]
```diff
@@ -236,7 +235,7 @@ public String endpointBody(final Request the_request, final Response the_respons
       }
      
       response_list.addAll(BallotSelection.toResponseList(cvr_to_audit_list));
-      response_list.sort(new BallotOrderComparator());
+      response_list.sort(null);
```
Comment (gws) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRToAuditDownload.java]: This looks weird but it's necessary - this is defined to use the natural ordering (`Comparable` implementations) if `null` is passed, but there is no 0-arity version of `sort`.

**Comment by gws** [comment]
This meets all of the CoAs in the referenced card - I'll be merging this into the epic branch later today regardless of review status (unless somebody finds a blocker). I'm just breaking up the PRs this way so folks get a feel for the changes.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/json/CountyDashboardRefreshResponse.java [hunk 2]
```diff
@@ -267,7 +269,7 @@ protected CountyDashboardRefreshResponse(final Long the_id,
     my_audited_ballot_count = the_audited_ballot_count;
     my_discrepancy_count = the_discrepancy_count;
     my_disagreement_count = the_disagreement_count;
-    my_ballot_under_audit_id = the_ballot_under_audit_id;
+    my_ballot_under_audit_ids = the_ballot_under_audit_ids;
```
Comment (sduckett) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/json/CountyDashboardRefreshResponse.java]: are we supposed to know which ballot under audit belongs to which audit board?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java [hunk 3]
```diff
@@ -536,6 +537,8 @@ public void endRound(final List<Elector> the_signatories) {
     if (my_current_round_index == null) {
       throw new IllegalStateException("no round to end");
     } else {
+      this.setAuditBoardCount(null);
```
Comment (sduckett) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java]: I guess `null` means we haven't picked a board size yet?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/json/CountyDashboardRefreshResponse.java (hunk 2 above)
Comment (gws) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/json/CountyDashboardRefreshResponse.java]: Yeah, if you look at how this is generated, it's a list of size `number_of_audit_boards` where each index in the list maps to an audit board, and there's a CVR ID for that audit board to look at next or `null` if the board has nothing left to do.

Comment (gws) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/json/CountyDashboardRefreshResponse.java]: It's up to the client to look at the right audit board index to get its next CVR (as well as looking itself up in the audit board signed-in data structure), that's the last card in the epic and what Sara and I are working on today.

File: server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java (hunk 3 above)
Comment (gws) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/model/CountyDashboard.java]: Right, the client knows about this and won't let audit boards sign in until there's some value here. Clearing it out when the round ends should suffice. I'm going to have to test more thoroughly to see if there are any edge cases - we're doing client work today and tomorrow.
